### PR TITLE
Fix TestDaemonRestartWithKilledRunningContainer failures on RHEL systems

### PR DESCRIPTION
**Pull request by mlaventure**
On RHEL system `TestDaemonRestartWithKilledRunningContainer` is currently failing due to a few races in containerd and docker.

This PR pulls in a new version of `containerd` fixing the found race (i.e. event were generated before their effect were reflected) and rework the restore code to ensure that we correctly get missed events for running containers during `docker` restore phase.

cc @tonistiigi @crosbymichael

### CODE REVIEW

**Comment by icecrime** [comment]
:grimacing: 

```
14:15:51 FAIL: docker_cli_run_test.go:2999: DockerSuite.TestRunPidHostWithChildIsKillable
14:15:51 
14:15:51 docker_cli_run_test.go:3018:
14:15:51     c.Fatal("Kill container timed out")
14:15:51 ... Error: Kill container timed out
```

**Comment by tonistiigi** [comment]
Panic in the experimental test:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x5f4c64]

goroutine 14 [running]:
panic(0x1b235a0, 0xc820014080)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/docker/docker/libcontainerd.(*client).Restore(0xc820016ae0, 0xc820012c80, 0x40, 0xc820015f20, 0x1, 0x1, 0x0, 0x0)
	/go/src/github.com/docker/docker/libcontainerd/client_linux.go:508 +0x324
github.com/docker/docker/plugin.(*Manager).restore(0xc82005c480, 0xc82007f680, 0x0, 0x0)
	/go/src/github.com/docker/docker/plugin/manager_linux.go:57 +0x31a
github.com/docker/docker/plugin.(*Manager).restorePlugin(0xc82005c480, 0xc82007f680, 0x0, 0x0)
	/go/src/github.com/docker/docker/plugin/manager.go:83 +0x142
github.com/docker/docker/plugin.(*Manager).init.func1(0xc820015f00, 0xc82005c480, 0xc82007f680)
	/go/src/github.com/docker/docker/plugin/manager.go:311 +0x81
created by github.com/docker/docker/plugin.(*Manager).init
	/go/src/github.com/docker/docker/plugin/manager.go:327 +0x353
```

**Comment by mlaventure** [comment]
all :green_heart:

File: libcontainerd/client_linux.go [hunk 1]
```diff
@@ -441,66 +435,130 @@ func (clnt *client) restore(cont *containerd.Container, options ...CreateOption)
...
 			}
-			return nil
+			logrus.Errorf("libcontainerd: failed to get container event for %s: %q", containerID, err)
+			return nil, err
+		}
+
+		logrus.Debugf("libcontainerd: received past event %#v", e)
+
+		switch e.Type {
+		case StateExit:
```
Comment (tonistiigi) [inline libcontainerd/client_linux.go]: nit: `case StateExit, StatePause, StateResume:`

File: libcontainerd/remote_linux.go [hunk 2]
```diff
@@ -283,45 +292,38 @@ func (r *remote) handleEventStream(events containerd.API_EventsClient) {
...
-				if err == nil {
-					break
-				}
-			}
-			r.RUnlock()
-			if container == nil {
-				logrus.Errorf("no state for container: %q", err)
-				continue
-			}
+		if e.Type == "live" {
```
Comment (tonistiigi) [inline libcontainerd/remote_linux.go]: Can't we just have `Stored: bool`. Last event uses `true` and this uses `false`.

Comment (mlaventure) [inline libcontainerd/remote_linux.go]: wdym?

Also just realized that I removed the generation of the "live" event from `containerd` by mistake, so I'll just delete this condition now.

**Comment by tonistiigi** [comment]
LGTM

**Comment by mlaventure** [comment]
All :green_heart: again :sweat_smile:

**Comment by icecrime** [comment]
LGTM :green_heart:

**Comment by crosbymichael** [comment]
LGTM

**Pull request merged**